src/form/df.jl

```diff
@@ -0,0 +1,163 @@
+    w_fr = pm.var[:nw][n][:w][f_bus]
+    w_to = pm.var[:nw][n][:w][t_bus]
+    ccm =  pm.var[:nw][n][:ccm][i]
+
+    #define series flow expressions to simplify Ohm's law
+    p_fr_s = p_fr - g_sh_fr*(w_fr/tm^2)
+    q_fr_s = q_fr + b_sh_fr*(w_fr/tm^2)
+
+    #KVL over the line:
+    @constraint(pm.model, w_to == (w_fr/tm^2) - 2*(r*p_fr_s + x*q_fr_s) + (r^2 + x^2)*ccm)
+
+end
+
+"""
+Defines relationship between branch (series) power flow, branch (series) current and node voltage magnitude
+"""
+function constraint_branch_current(pm::GenericPowerModel{T}, n::Int, i, f_bus, f_idx, g_sh_fr, b_sh_fr, tm) where T <: AbstractDFForm
+    p_fr = pm.var[:nw][n][:p][f_idx]
+    q_fr = pm.var[:nw][n][:q][f_idx]
+    w_fr = pm.var[:nw][n][:w][f_bus]
+    ccm  = pm.var[:nw][n][:ccm][i]
+
+    #define series flow expressions to simplify constraint
+    p_fr_s = p_fr - g_sh_fr*(w_fr/tm^2)
+    q_fr_s = q_fr + b_sh_fr*(w_fr/tm^2)
+
+    #convex constraint linking p, q, w and ccm
+    @constraint(pm.model, p_fr_s^2 +q_fr_s^2 <= (w_fr/tm^2)*ccm)
+end
+
+
+function constraint_voltage_angle_difference(pm::GenericPowerModel{T}, n::Int, arc_from, f_bus, t_bus, angmin, angmax) where T <: AbstractDFForm
+    # how to evolve the constraint template? - new method definition only for this formulation?
+    i = arc_from[1]
+    f_idx = arc_from
+    t_idx = (i, t_bus, f_bus)
+
+    branch = ref(pm, n, :branch, i)
+    c = branch["br_b"]
+    tm = branch["tap"]
+    g, b = calc_branch_y(branch)
+    tr, ti = calc_branch_t(branch)
+
+    # convert series admittance to impedance
+    z_s = 1/(g + im*b)
+    r_s = real(z_s)
+    x_s = imag(z_s)
+
+    # to support asymmetric shunts + conductance
+    g_sh_fr = 0
+    g_sh_to = 0
+    b_sh_fr = c/2
+    b_sh_to = c/2
+
+    # getting the variables
+    w_fr = pm.var[:nw][n][:w][f_bus]
+    w_to = pm.var[:nw][n][:w][t_bus]
+
+    p_fr = pm.var[:nw][n][:p][f_idx]
+    p_to = pm.var[:nw][n][:p][t_idx]
+
+    q_fr = pm.var[:nw][n][:q][f_idx]
+    q_to = pm.var[:nw][n][:q][t_idx]
+
+    #TODO: adapt for asymmetric shunts + shunt conductance
+    tzr = r_s*tr - x_s*ti
+    tzi = r_s*ti + x_s*tr
+    # a1 = (( tr - tzi*b_sh_fr)*(w_fr/tm^2) - tzr*p_fr - tzi*q_fr)
+    # a2 = ((-ti - tzr*b_sh_fr)*(w_fr/tm^2) - tzr*q_fr + tzi*p_fr)
+    # a3 = (( tr - tzi*b_sh_fr)*(w_fr/tm^2) - tzr*p_fr - tzi*q_fr)
+    # a4 = ((-ti - tzr*b_sh_fr)*(w_fr/tm^2) - tzr*q_fr + tzi*p_fr)
+    @constraint(pm.model,
+        tan(angmin)*(( tr - tzi*b_sh_fr)*(w_fr/tm^2) - tzr*p_fr - tzi*q_fr)
+                 <= ((-ti - tzr*b_sh_fr)*(w_fr/tm^2) - tzr*q_fr + tzi*p_fr)
+        )
+    @constraint(pm.model,
+        tan(angmax)*(( tr - tzi*b_sh_fr)*(w_fr/tm^2) - tzr*p_fr - tzi*q_fr)
+                 >= ((-ti - tzr*b_sh_fr)*(w_fr/tm^2) - tzr*q_fr + tzi*p_fr)
+        )
+
+end
+
+
+"variable: `0 <= i[l] <= (Imax)^2` for `l` in `branch`es"
+function variable_branch_series_current_magnitude_sqr(pm::GenericPowerModel, n::Int=pm.cnw; bounded = true)
```

Comment:
Can we replace this function with the one that is already defined here, https://github.com/lanl-ansi/PowerModels.jl/blob/master/src/form/wr.jl#L399?  If so, that function can move to core/variable.jl because it is generally applicable.

Reply:
These constraints have different meaning: bounding series current in this BFM derivation vs bounding total current magnitude in QC. Anyway, I updated the series current upper bound with a compensation term for the current contribution of the shunts.